### PR TITLE
finish adding `ref` and `selector` fields for all Grafana OnCall resources

### LINKED ISSUE #4411: Crossplane - add missing selector and ref fields for various OnCall resources
### What would you like to see!

Our Crossplane provider currently doesn't allow specifying a selector for an integration's default route's escalation chain (we only provide `defaultRoute.escalationChainId`). See [our crossplane provider docs](https://marketplace.upbound.io/providers/grafana/provider-grafana/v0.16.1/resources/oncall.grafana.crossplane.io/Integration/v1alpha1#doc:spec-initProvider-defaultRoute-escalationChainId) for more context.

See [`NotificationPolicy.contactPointRef`](https://marketplace.upbound.io/providers/grafana/provider-grafana/v0.16.1/resources/alerting.grafana.crossplane.io/NotificationPolicy/v1alpha1#doc:spec-forProvider-contactPointRef) as an example of ideally how we should do this.

**Deliverables**
As a user of Crossplane whom is trying to code-ify my OnCall Integrations, I should be able to have my Integration's default route escalation chain reference an Escalation Chain

### Product Area

Terraform

### Anything else to add?

_No response_

### PR DESCRIPTION
**Pull request by joeyorlando**
### Description of your changes

Adds all missing `selector` and `ref` fields for the following Grafana OnCall Crossplane resources:
- `oncall.grafana.crossplane.io.EscalationChain`
- `oncall.grafana.crossplane.io.Escalation`
- `oncall.grafana.crossplane.io.Integration`
- `oncall.grafana.crossplane.io.OnCallShift`
- `oncall.grafana.crossplane.io.OutgoingWebhook`
- `oncall.grafana.crossplane.io.Route`
- `oncall.grafana.crossplane.io.Schedule`

Closes https://github.com/grafana/oncall/issues/4411

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### CODE REVIEW

**Comment by joeyorlando** [comment]
@julienduchesne I'm trying to run `make reviewable test` locally (as per the PR description template) and getting:
```bash
❯ make reviewable test
13:08:43 [ .. ] generating provider schema for grafana/grafana 2.19.1
make[2]: *** [config/schema.json] Error 1
make[1]: *** [generate] Error 2
make: *** [reviewable] Error 2
```

**Comment by joeyorlando** [comment]
@MasterExecuson see this PR.

One thing to note is that for any related object which is only represented by a Terraform datasource (as opposed to a resource), Crossplane doesn't allow us to specify refs/selectors for these. Terraform datasources are not translated to Crossplane resources unfortunately.

The workaround they have is [this](https://github.com/crossplane/crossplane/blob/master/design/design-doc-observe-only-resources.md), basically using the Terraform provider for Crossplane to use the datasources directly.

Once we get v0.17.0 of this provider published, let me know if you're able to get this working with the linked documentation.

File: config/grafana/config.go [hunk 1]
```diff
@@ -28,6 +28,101 @@ func ConfigureOrgIDRefs(p *ujconfig.Provider) {
...
+
+	p.AddResourceConfigurator("grafana_oncall_route", func(r *ujconfig.Resource) {
+		r.References["escalation_chain_id"] = escalationChainRef
+		r.References["integration_id"] = integrationRef
+	})
+
+	p.AddResourceConfigurator("grafana_oncall_schedule", func(r *ujconfig.Resource) {
+		// TODO: this should be a list.. will this work like this?
+		// for reference https://registry.terraform.io/providers/grafana/grafana/latest/docs/resources/oncall_schedule#shifts
+		r.References["shifts"] = oncallShiftRef
```
Comment (julienduchesne) [inline config/grafana/config.go]: For fields with special names like this, I wouldn't call it `OnCallShiftRef`. It'd call it `ShiftsRef` or something, so that it's clear to which field it's mapped

As it is, the generated CRD looks like this:
```
...
                  onCallShiftRef:
                    description: References to OnCallShift in oncall to populate shifts.
                    items:
                      description: A Reference to a named object.
                      properties:
                        name:
                          description: Name of the referenced object.
                          type: string
                        policy:
                          description: Policies for referencing.
                          properties:
                            resolution:
                              default: Required
                              description: |-
                                Resolution specifies whether resolution of this reference is required.
                                The default is 'Required', which means the reconcile will fail if the
                                reference cannot be resolved. 'Optional' means this reference will be
                                a no-op if it cannot be resolved.
                              enum:
                              - Required
                              - Optional
                              type: string
                            resolve:
                              description: |-
                                Resolve specifies when this reference should be resolved. The default
                                is 'IfNotPresent', which will attempt to resolve the reference only when
                                the corresponding field is not present. Use 'Always' to resolve the
                                reference on every reconcile.
                              enum:
                              - Always
                              - IfNotPresent
                              type: string
                          type: object
                      required:
                      - name
                      type: object
                    type: array
                  onCallShiftSelector:
                    description: Selector for a list of OnCallShift in oncall to populate
                      shifts.
                    properties:
                      matchControllerRef:
                        description: |-
                          MatchControllerRef ensures an object with the same controller reference
                          as the selecting object is selected.
                        type: boolean
                      matchLabels:
                        additionalProperties:
                          type: string
                        description: MatchLabels ensures an object with matching labels
                          is selected.
                        type: object
                      policy:
                        description: Policies for selection.
                        properties:
                          resolution:
                            default: Required
                            description: |-
                              Resolution specifies whether resolution of this reference is required.
                              The default is 'Required', which means the reconcile will fail if the
                              reference cannot be resolved. 'Optional' means this reference will be
                              a no-op if it cannot be resolved.
                            enum:
                            - Required
                            - Optional
                            type: string
                          resolve:
                            description: |-
                              Resolve specifies when this reference should be resolved. The default
                              is 'IfNotPresent', which will attempt to resolve the reference only when
                              the corresponding field is not present. Use 'Always' to resolve the
                              reference on every reconcile.
                            enum:
                            - Always
                            - IfNotPresent
                            type: string
                        type: object
                    type: object
                  shifts:
                    description: |-
                      call shifts.
                      The list of ID's of on-call shifts.
                    items:
                      type: string
                    type: array
                    x-kubernetes-list-type: set
...
```

Comment (joeyorlando) [reply config/grafana/config.go]: thanks for that @julienduchesne !

File: config/grafana/config.go [hunk 2]
```diff
@@ -28,6 +28,98 @@ func ConfigureOrgIDRefs(p *ujconfig.Provider) {
...
+	scheduleRef := ujconfig.Reference{
+		TerraformName:     "grafana_oncall_schedule",
+		RefFieldName:      "ScheduleRef",
+		SelectorFieldName: "ScheduleSelector",
+	}
+
+	p.AddResourceConfigurator("grafana_oncall_escalation", func(r *ujconfig.Resource) {
+		r.References["escalation_chain_id"] = escalationChainRef
+		r.References["action_to_trigger"] = outgoingWebhookRef
+		r.References["notify_on_call_from_schedule"] = scheduleRef
```
Comment (julienduchesne) [inline config/grafana/config.go]: These two also, I would give them different names

- ActionToTriggerRef
- NotifyOnCallFromScheduleRef

File: package/crossplane.yaml
Comment (joeyorlando) [inline package/crossplane.yaml:1]: see [this conversation](https://raintank-corp.slack.com/archives/C03A2L4R9U1/p1718723636803719?thread_ts=1718722993.484269&cid=C03A2L4R9U1)

tldr; I had to manually make this change to be able to create this `Provider` resource locally (`meta...` apiVersion didn't exist)

**Review by julienduchesne — APPROVED** [review]
LGTM!
